src/functional/otf/compilation/build_systems/cmake.py

```diff
@@ -0,0 +1,145 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import annotations
+
+import dataclasses
+import pathlib
+import subprocess
+from typing import Optional
+
+from functional.otf import languages, stages
+from functional.otf.compilation import build_data, cache, common, compiler
+from functional.otf.compilation.build_systems import cmake_lists
+
+
+@dataclasses.dataclass
+class CMakeFactory(
```

Comment:
I had them as functions. It was incredibly ugly and much harder to read.

Reply:
To clarify: I had them as functions, which returned functions. The compiling step would then call the inner function, so that it doesn't have to know about build system specific config. Hence it was an uglier and harder to read version of this and the config was still stored in the inner function's closure. Implementing this as one function would couple the compiling step to each and every build system.

Reply:
Also conceptually you would then have a build system builder configurator (builder builder?, maker builder?), which makes for really ugly names.

Reply:
What I meant was:
```python
def cmake_factory(
    source: stages.CompilableSource[...],
    cache_strategy: cache.Strategy,
    cmake_generator_name: str = "Ninja"
    cmake_build_type: str = "Debug"
    cmake_extra_flags: Optional[list[str]] = None
):
  ...
```
That is, merge the parameter list of `__init__` and `__call__`, and have a single function. But now I understand that you need to separate the build-system-specific information from the program-specific information, in which case I think it's good the way it is.